[10.x] Add Model::isValidColumn method

calebdw opened the pull request:
<!--
Please only send a pull request to branches that are currently supported: https://laravel.com/docs/releases#support-policy 

If you are unsure which branch your pull request should be sent to, please read: https://laravel.com/docs/contributions#which-branch

Pull requests without a descriptive title, thorough description, or tests will be closed.

In addition, please describe the benefit to end users; the reasons it does not break any existing features; how it makes building web applications easier, etc.
-->

Hello!

This PR introduces a new `Model::isValidColumn` method that checks if the given column is a valid database column.

### Why
I need to check if a column is a valid column in the database---there's currently not an easy, efficient way to do this.

You can use the `Schema` facade:
```php
Schema::hasColumn($model->getTable(), $column);
```
however, this performs a query for every single check---not good for loops.

The next best solution is to statically cache all the table columns:
```php
protected static array $tableColumns;
//...
if (! isset(static::$tableColumns)) {
    static::$tableColumns = Schema::getColumnListing($model->getTable());
}
//...
if (in_array($column, static::$tableColumns)) {
    //...
}
```

Which works, but then I discovered that the Model is **already doing this** via `GuardsAttributes::isGuardableColumn`---it's just a protected method that I don't have direct access to:

https://github.com/laravel/framework/blob/ff96905683e7035d6d7c9700ef8f722344c27c2f/src/Illuminate/Database/Eloquent/Concerns/GuardsAttributes.php#L215-L229

To keep the logic on the Model and avoid duplication/BCs I moved the logic from `GuardsAttributes::isGuardableColumn` to `Model::isValidColumn` (and update the old location to point to the new logic).

### Alternatives
- Can change the visibility of `GuardsAttributes::isGuardableColumn` from `protected` to `public`, however, 1) I wasn't sure if this would be considered a BC, and 2) it's not the most descriptive in the context that I want to use the method---checking if it's a **valid database** column not just a *guardable* column.


Thanks!

calebdw committed:
Add Model::isValidColumn method

taylorotwell commented:
Can you define a trait in your own project with this method name and simply call the protected method from there?

CalebDW commented:
@taylorotwell, sure I could do that...but would need to create an AbstractModel that all models extend. But is this not something that could benefit the community at large? I can't be the only one that needs to ensure that an attribute is a valid database column.

donnysim commented:
Could you describe a situation where this would be needed? Because checking table schema is not really an efficient solution because you still need to establish a connection where none could be needed. 

From my experience if you're creating the application you should know if the column exists or not, the situation where you would check columns individually from database is very rare and I doubt many in that case would resolve to checking the database schema for it, mostly would go around using feature flags or such.

CalebDW commented:
@donnysim,

We have dedicated query builder classes, and we've added additional logic to more easily apply request parameters to scopes. One of the conveniences I've added is for simple `where(<column>, <value>)`  or `orderBy(<column>)` clauses, a method does not have to be explicitly defined on the Builder class---instead I check if the attribute is a valid database column and if so then I apply the `where`/`orderBy`. I need to check that it's a valid column to prevent SQL exceptions if an invalid column were to be passed in

donnysim commented:
But if that's the case wouldn't you also want to validate the data being passed from request parameters which would require specifying the column validation rules anyway so that it doesn't end up like `user_id = "1,2,3"` and you get a incorrect value, truncating warning and so on?

CalebDW commented:
Take this example:
I allow sorting to be dynamically specified with a `sort_by` attribute in the form of `'-completed_at,id'` (conforms to JSON:API specs). I then loop over all the columns and determine if they should be sorted asc/desc. However, now I need to validate that those are actual DB columns or the `orderBy` will throw an exception.

There's no reason for the class to throw a SQL exception that alerts all devs when I can just skip it.

dennisprudlo commented:
Valid use case, however I think I would be more explicit and define a `$sortable` property with the column names that can be sorted (even if I have to put all column names in there). Yes, it is less generic and it wouldn't just work when adding a new column to the database – you would have to add the column name to the property – but you get rid of that extra database query to check which columns actually exist.

We built an app for a client once where they needed a form builder and every form should be a table in a specific database. So tables where added and altered automatically (when a new field was added for example). Upon saving the form after editing it we had to sync the table (check if a newly added "first_name" field already exists in the table and so on).

But I don't know if this is really required as a method in the model core as @donnysim said:
> the situation where you would check columns individually from database is very rare

donnysim commented:
> @dennisprudlo
> 
> > but you get rid of that extra database query to check which columns actually exist.
> 
> The thing is, _all models are already performing this query and caching the result_

where did you get that? Laravel models do nothing to know about guarded fields by database. Every attribute that is not in $fillable is by default guarded no matter if it exists or not in database. Or are you referencing your project? I personally always disable the guarded property because it's better to be explicit what you're filling and I've seen many cases where developers add an x_id to fillable (and I did in my early days too) just because it didn't set, instead of understanding the problem.

CalebDW commented:
@donnysim, 

> Laravel models do nothing to know about guarded fields by database. 

You are incorrect, see `GuardsAttributes::isGuardableColumn()` which is called by `GuardsAttributes::isGuarded()` which in turn is called by `GuardsAttributes::isFillable()`.

donnysim commented:
Oh, never knew that 😆to be honest I'm even more glad I disable them now.

dennisprudlo commented:
@CalebDW This is very interesting. Do you know why the database columns are being read? I mean as you said the `GuardsAttributes::isGuardableColumn` is basically a `GuardsAttributes::isValidColumn` and only used in `GuardsAttributes::isGuarded`:

```php
public function isGuarded($key)
{
    if (empty($this->getGuarded())) {
        return false;
    }

    return $this->getGuarded() == ['*'] ||
           ! empty(preg_grep('/^'.preg_quote($key, '/').'$/i', $this->getGuarded())) ||
           ! $this->isGuardableColumn($key);
}
```

Say we have given a column for a mass update with the name "foo". Now the code says this column is guarded, when
- The `$guarded` property is `['*']`, meaning all columns should be guarded
- The value "foo" is somewhere in the `$guarded` array, meaning it was explicitly defined as guarded
- The value "foo" is not a valid column.

But why the last one? Why not simply let the database crash with a "column not found" when you give it a nonexisten column. I mean ultimately it is the developers fault and laravel doesn't do that in other cases. Like you get exceptions when you didn't hook up a relationship the correct way instead of `null` or when using `MyModel::get()` but the defined table does not exist – instead of an empty array.

This would obviously cost database queries to find out in the first place but why is it acceptable in the guarded case? Anyone knows why we do that here?
